src/compiler/scala/tools/nsc/backend/jvm/BCodeAsmCommon.scala

```diff
@@ -300,4 +302,73 @@ final class BCodeAsmCommon[G <: Global](val global: G) {
+   * captured when creating the completer in the namer. However, we can temporarily replace
+   * global.reporter (it's a var) to store errors.
+   */
+  def completeSilentlyAndCheckErroneous(sym: Symbol): Boolean = {
+    if (sym.hasCompleteInfo) false
+    else {
+      val originalReporter = global.reporter
+      val storeReporter = new reporters.StoreReporter()
+      global.reporter = storeReporter
+      try {
+        sym.info
+      } finally {
+        global.reporter = originalReporter
+      }
+      sym.isErroneous
+    }
+  }
+
+  /**
+   * Build the [[InlineInfo]] for a class symbol.
+   */
+  def buildInlineInfoFromClassSymbol(classSym: Symbol, classSymToInternalName: Symbol => InternalName, methodSymToDescriptor: Symbol => String): InlineInfo = {
+    val selfType = {
+      // The mixin phase uses typeOfThis for the self parameter in implementation class methods.
+      val selfSym = classSym.typeOfThis.typeSymbol
+      if (selfSym != classSym) Some(classSymToInternalName(selfSym)) else None
+    }
+
+    val isEffectivelyFinal = classSym.isEffectivelyFinal
+
+    // Primitive methods cannot be inlined, so there's no point in building a MethodInlineInfo. Also, some
+    // primitive methods (e.g., `isInstanceOf`) have non-erased types, which confuses [[typeToBType]].
+    val methodInlineInfos = classSym.info.decls.iterator.filter(m => m.isMethod && !scalaPrimitives.isPrimitive(m)).flatMap({
+      case methodSym =>
+        if (completeSilentlyAndCheckErroneous(methodSym)) {
+          // Happens due to SI-9111. Just don't provide any MethodInlineInfo for that method, we don't need fail the compiler.
+          if (!classSym.isJavaDefined) devWarning("SI-9111 should only be possible for Java classes")
+          None
+        } else {
+          val name      = methodSym.javaSimpleName.toString // same as in genDefDef
+          val signature = name + methodSymToDescriptor(methodSym)
+
+          // Some detours are required here because of changing flags (lateDEFERRED, lateMODULE):
+          // 1. Why the phase travel? Concrete trait methods obtain the lateDEFERRED flag in Mixin.
+          //    This makes isEffectivelyFinalOrNotOverridden false, which would prevent non-final
+          //    but non-overridden methods of sealed traits from being inlined.
+          // 2. Why the special case for `classSym.isImplClass`? Impl class symbols obtain the
+          //    lateMODULE flag during Mixin. During the phase travel to exitingPickler, the late
+          //    flag is ignored. The members are therefore not isEffectivelyFinal (their owner
+          //    is not a module). Since we know that all impl class members are static, we can
+          //    just take the shortcut.
+          val effectivelyFinal = classSym.isImplClass || exitingPickler(methodSym.isEffectivelyFinalOrNotOverridden)
+          val isTraitMethodNonAccessor = (classSym.isTrait && !classSym.isImplClass) && !methodSym.isAccessor
```

Comment:
```
tail -100 sandbox/Protected.java sandbox/test.scala
==> sandbox/Protected.java <==
package other;

public class Protected {
	protected String foo() { return "foo"; }
}

==> sandbox/test.scala <==
class C extends other.Protected {
  class Inner {
    def bar = foo
  }
}
object Test {
  def main(args: Array[String]): Unit = {
    val c = new C {}
    val inner = new c.Inner
    println(inner.bar)
  }
}
2.11.x ~/code/scala2 javac -d . sandbox/Protected.java && qscalac -Xprint:superaccessors sandbox/test.scala && qscala Test
[[syntax trees at end of            superaccessors]] // test.scala
package <empty> {
  class C extends other.Protected {
    <artifact> def protected$foo(x$1: C)(): String = x$1.foo();
    def <init>(): C = {
      C.super.<init>();
      ()
    };
    class Inner extends scala.AnyRef {
      def <init>(): C.this.Inner = {
        Inner.super.<init>();
        ()
      };
      def bar: String = C.this.protected$foo(C.this)()
    }
  };
  object Test extends scala.AnyRef {
    def <init>(): Test.type = {
      Test.super.<init>();
      ()
    };
    def main(args: Array[String]): Unit = {
      val c: C = {
        final class $anon extends C {
          def <init>(): <$anon: C> = {
            $anon.super.<init>();
            ()
          }
        };
        new $anon()
      };
      val inner: c.Inner = new c.Inner();
      scala.this.Predef.println(inner.bar)
    }
  }
}

foo
```

We currently have implementation restrictions the prevent this from working if `C` were a trait, so it might not matter for this case. But you could future proof by handling this.